oauthproxy.go

```diff
@@ -838,6 +839,27 @@ func (p *OAuthProxy) enrichSessionState(ctx context.Context, s *sessionsapi.Sess
 func (p *OAuthProxy) AuthOnly(rw http.ResponseWriter, req *http.Request) {
 	session, err := p.getAuthenticatedSession(rw, req)
 	if err != nil {
+		if authOnlyRedirectSignIn(req) {
+			if isAjax(req) {
+				// no point redirecting an AJAX request
+				p.errorJSON(rw, http.StatusUnauthorized)
+				return
+			}
+
+			// Orignal request URI is captured and added using `rd` parameter in the redirect
+			// to ensure a redirect to the original URL will still happen after OAuth/Signin process
+			redirect, err := p.getAppRedirect(req)
+			if err != nil {
+				logger.Errorf("Error obtaining redirect: %v", err)
+				p.ErrorPage(rw, http.StatusInternalServerError, "Internal Server Error", err.Error())
+				return
+			}
+
+			// SignIn has the check on SkipProviderButton to do either SigninPage or OauthStart
+			redirectPath := fmt.Sprintf("%s?rd=%s", p.SignInPath, redirect)
+			http.Redirect(rw, req, redirectPath, http.StatusFound)
```

Comment:
Do you need the full URL here?

E.g. -- what happens if this is used in a system hitting this from a subrequest auth check where the OAuth2-Proxy server is hosted on a different subdomain then the requested resource.

In that case, the redirect to the relative path will be at the application domain - not the separate oauth2-proxy server. (this separate central oauth2-proxy auth server hit via subrequests on a different subdomain is a common architecture deployment).

Reply:
And then how do we accurately know the appropriate `Proto/Host` details to set? If this endpoint is hit via subrequests, it might not have `X-Forwarded` details or be on different internal domain (and therefore different HTTP Host header).

For instance, when I use this architecture, I do subrequests to this endpoint over the Kubernetes service URL directly (e.g. `deployment.namespace.svc.cluster.local`) to avoid the extra hop of going though the Ingress layer internally.

The views the `/oauth2/auth` endpoint has of its `requestutil.GetRequestHost(req)` for those requests are not what is accessible externally.

Reply:
Upon further reflection -- this problem will cause you to get misleading results for the `p.getAppRedirect(req)` call above to try to set an `rd` querystring as well 🤔

Reply:
I think we might need to revisit `redirect_signin` not being a bool, but rather being an explicit URL to the sign in endpoint with the proto/host/path all set in your configuration.

Reply:
Apologies but I quite don't understand by what you mean by `subrequests` but perhaps an example would help me. However, interesting that you mention bypassing ingress and the redirect logic depends upon ingress configuration.

> For instance, when I use this architecture, I do subrequests to this endpoint over the Kubernetes service URL directly (e.g. deployment.namespace.svc.cluster.local) to avoid the extra hop of going though the Ingress layer internally

Trying to understand how this looks for real requests, an end to end example of what the initial request is and how you bypass the ingress and make the `/oauth2/auth` request to the kubernetes internal endpoints would be useful. We do have to note that since `redirect_signin` is a boolean flag (for now), you could still switch off this behaviour by setting it to false when using internal endpoints?

> I think we might need to revisit `redirect_signin` not being a bool, but rather being an explicit URL to the sign in endpoint with the proto/host/path all set in your configuration.
I possibly didn't get your point correctly but any explicit configuration will not be possible to set when using ingress like Traefik